drivers/io-expander/demux_spi/demux_spi.c

```diff
@@ -0,0 +1,155 @@
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include "demux_spi.h"
+#include "error.h"
+#include <stdlib.h>
+#include <string.h>
+
+/******************************************************************************/
+/**************************** Types Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Demux specific SPI platform ops structure
+ */
+const struct spi_platform_ops demux_spi_platform_ops = {
+	.spi_ops_init = demux_spi_init,
+	.spi_ops_remove = demux_spi_remove,
+	.spi_ops_write_and_read = demux_spi_write_and_read
+};
+
+/**
+ * @brief Initialize the SPI demux layer.
+ * @param desc - The SPI descriptor.
+ * @param param - The structure that contains the SPI parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t demux_spi_init(struct spi_desc **desc,
+		       const struct spi_init_param *param)
+{
+	int32_t ret;
+
+	struct spi_desc *descriptor;
+	struct spi_desc *spi_dev_desc;
+	struct spi_init_param *spi_dev_param;
+
+	if(!param)
+		return FAILURE;
+
+	descriptor = (struct spi_desc *)calloc(1, sizeof(*descriptor));
+	if(!descriptor) {
+		free(descriptor);
+		return FAILURE;
+	}
+
+	descriptor->extra = (struct spi_desc *)calloc(1, sizeof(*(descriptor->extra)));
+	if(!(descriptor->extra)) {
+		free(descriptor->extra);
+		return FAILURE;
+	}
+
+	descriptor->chip_select = param->chip_select;
+	descriptor->max_speed_hz = param->max_speed_hz;
+	descriptor->mode = param->mode;
+
+	spi_dev_param = param->extra;
+
+	ret = spi_init(&spi_dev_desc, spi_dev_param);
+
+	(descriptor->extra) = spi_dev_desc;
+
+	*desc = descriptor;
+
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by demux_spi_init().
+ * @param desc - The SPI descriptor.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t demux_spi_remove(struct spi_desc *desc)
+{
+	if(spi_remove(desc->extra))
+		return FAILURE;
+	free(desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Write and read data to/from SPI demux layer.
+ * @param desc - The SPI descriptor.
+ * @param data - The buffer with the transmitted/received data.
+ * @param bytes_number - Number of bytes to write/read.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t demux_spi_write_and_read(struct spi_desc *desc, uint8_t *data,
+				 uint16_t bytes_number)
+{
+	int32_t ret;
+	uint8_t cs;
+	uint8_t *buff;
+
+	struct spi_desc *spi_dev;
+
+	buff = malloc(sizeof(*buff) * (bytes_number+1));
+	if(!buff)
+		return FAILURE;
+
+	spi_dev = desc->extra;
+	cs = 0x80 | desc->chip_select;
+
+	buff[0] = cs;
+	memcpy((buff+1), data, bytes_number);
+
+	ret = spi_write_and_read(spi_dev, buff, bytes_number+1);
```

Comment:
if ret is negative, what value does bytes_number have? `memcpy` might do some things here.